### PR TITLE
Prompt to sign in to GitHub hosts in credential helper

### PR DESCRIPTION
**Pull request by niik**
## Description

Follow-up to #18700. This PR adds support for prompting users to sign in to GitHub hosts in the credential helper. In order to accomplish this we need a way to abstract the entire sign-in state machine in one operation (i.e. a promise).

This moves the OAuth dance into the sign-in store and creates a promise that encompasses the entire sign-in flow.

Further more this works around the issue in #18700 where we would show the user an error saying 'fatal: could not read Username for '': terminal prompts disabled' when we couldn't find credentials. We now detect that case and throw the good old fashioned "Authentication failed" error instead.

## Release notes

Notes:

### CODE REVIEW

**Comment by niik** [comment]
@sergiou87 @tidy-dev I feel like this should be a required piece of #18700 so I'd love it if we could review it here and merge it into #18700 and then land the entire piece on development

File: app/src/lib/api.ts [hunk 1]
```diff
@@ -2212,3 +2218,75 @@ function tryUpdateEndpointVersionFromResponse(
...
+    return true
+  }
+
+  if (isKnownThirdPartyHost(hostname)) {
+    return false
+  }
+
+  // bitbucket.example.com, etc
+  if (/(^|\.)(bitbucket|gitlab)\./.test(hostname)) {
+    return true
```
Comment (sergiou87) [inline app/src/lib/api.ts]: Shouldn't this be false? What am I missing?

File: app/src/lib/git/credential.ts [hunk 2]
```diff
@@ -1,13 +1,41 @@
...
+  // and expand multiple key[] entries into a key[0], key[1]... key[n] sequence.
+  // We then remove the number from the key when we're formatting the credential
+  for (const [, k, v] of value.matchAll(/^(.*?)=(.*)$/gm)) {
+    if (k.endsWith('[]')) {
+      let i = 0
+      const newKey = `${k.slice(0, -2)}[${i}]`
+      while (cred.has(newKey)) {
+        i++
+      }
+      cred.set(newKey, v)
```
Comment (sergiou87) [inline app/src/lib/git/credential.ts]: Am I missing something here? `newKey` never changes 🤔 

Could you add to the comment a sample input for `parseCredential` and then the expected output for `formatCredential` for that same info?

File: app/src/lib/trampoline/trampoline-credential-helper.ts [hunk 3]
```diff
@@ -90,67 +93,114 @@ async function getExternalCredential(input: Credential, token: string) {
 /** Implementation of the 'get' git credential helper command */
 async function getCredential(cred: Credential, store: Store, token: string) {
   const ghCred = await getGitHubCredential(cred, store)
+  const accounts = await store.getAll()
 
-  if (ghCred || (await isCredentialStoredInternally(cred, store))) {
+  if (ghCred) {
     return ghCred
   }
```
Comment (sergiou87) [inline app/src/lib/trampoline/trampoline-credential-helper.ts]: I guess we can get the accounts after the `if (ghCred)`?

**Review by sergiou87 — CHANGES_REQUESTED** [review]
Looks good overall, only found a few things

File: app/src/lib/api.ts (hunk 1 above)
Comment (niik) [reply app/src/lib/api.ts]: You're not missing anything, last minute addition by a sleep deprived mind. I'll fix this!

File: app/src/lib/git/credential.ts (hunk 2 above)
Comment (niik) [reply app/src/lib/git/credential.ts]: yeah 🤦 I should stop doing last minute cleanups

Comment (niik) [reply app/src/lib/git/credential.ts]: I added two tests that hopefully should explain this clearly

Comment (sergiou87) [reply app/src/lib/git/credential.ts]: Those tests are perfect, thank you!

**Review by sergiou87 — APPROVED** [review]
Thank you for those changes! I've been testing this and it works great 😍

**Comment by niik** [comment]
Thank you @sergiou87!

**Pull request merged**